src/main/java/com/google/sps/servlets/utils/Statistics.java

```diff
@@ -54,43 +63,41 @@ public Statistics(List<Double> sentimentScores) {
    */
   private void setAggregateScores(List<Double> sentimentScores) {
     aggregateValues = new HashMap<>();
-    BigDecimal curPoint = BigDecimal.valueOf(LOWER_END);
-    BigDecimal interval = BigDecimal.valueOf(INTERVAL);
-
-    // Initialize the HashMap with intervals
-    while (curPoint.doubleValue() < UPPER_END) {
-      Range currentRange =
-          new Range(
-              curPoint.doubleValue(), Math.min(curPoint.add(interval).doubleValue(), UPPER_END));
-      curPoint = curPoint.add(interval);
-      aggregateValues.put(currentRange, 0);
-    }
     // Add score's interval to different ranges two sorting with and two pointers pop-up
     sentimentScores.sort(Comparator.naturalOrder());
-    curPoint = BigDecimal.valueOf(LOWER_END);
-    int scoreIdx = 0;
-    while ((scoreIdx < sentimentScores.size()) && (curPoint.doubleValue() < UPPER_END)) {
-      double scoreVal = sentimentScores.get(scoreIdx);
-      // check available interval for scoreVal
-      BigDecimal nextPoint = curPoint.add(interval);
-      double intervalUpperVal = nextPoint.doubleValue();
-      if (scoreVal < intervalUpperVal || intervalUpperVal == UPPER_END) {
-        Range foundRange = new Range(curPoint.doubleValue(), intervalUpperVal);
-        aggregateValues.put(foundRange, aggregateValues.get(foundRange) + 1);
-        scoreIdx += 1;
-      } else {
-        curPoint = nextPoint;
+    int updatingScoreIdx = 0;
+    for (BigDecimal tempPoint = LOWER_END;
+        tempPoint.compareTo(UPPER_END) < 0;
+        tempPoint = tempPoint.add(INTERVAL)) {
+      BigDecimal nextPoint = UPPER_END.min(tempPoint.add(INTERVAL));
+      Range currentRange = new Range(tempPoint, nextPoint);
+      aggregateValues.put(currentRange, 0);
+      for (int scoreIdx = updatingScoreIdx; scoreIdx < sentimentScores.size(); scoreIdx++) {
+        BigDecimal scorePoint = BigDecimal.valueOf(sentimentScores.get(scoreIdx));
+        if ((scorePoint.compareTo(nextPoint) < 0) || nextPoint.compareTo(UPPER_END) == 0) {
```

Comment:
This algorithm now looks pretty complex, given it does many things. Can you add a couple comments to explain what it is doing?